### DISCUSSION TITLE
Quarto book preview: "WARN: connection closed before message completed" any clue?

### DISCUSSION THREAD

**Discussion by simosacchi** [quarto-dev/quarto-cli#9933] [Q&A]
### Description

Dear all,

I am experiencing a frustrating situation.  I am on MacOS Sonoma 14.4.1, Quarto 1.5.45, and the same behaviour happens with Chrome, Sarafi and Firefox.
This is the issue. After a while (and I cannot reproduce exactly when), when I launch "quarto preview" (options do not change the behaviour), I start getting a "WARN: connection closed before message completed" when navigating the website, and I am stuck on the original page. The behaviour seems to hit after several pages are added to the YAML file. All pages are just Markdown (no code).

I have tried building the site from scratch several times, the last one keeping the preview running with the --watch-inputs. It worked properly until I killed the process. Running the process again brought back the warning and got me stuck again.

Even more strangely, repeatedly hitting the link gets me to the desired page at some point. Moreover, if I open different pages of the website (e.g. http://localhost:3604/index.html and http://localhost:3604/preface.html) on different tabs, once I reach three open pages, I can see in the STDOUT of the terminal that they reload in a cycle without stopping.

I am completely puzzled.

Thank you for any suggestions.

best,
Simone

**Comment by simosacchi** [comment]
I have just discovered that "quarto preview --render html" creates yet another scenario:
- if no changes are applied, the preview works properly
- a change to the YAML file is enough to bring the WARN back.

**Reply by mcanouil** [reply]
Without a fully reproducible example/workflow, hard to guess or figure out the issue unfortunately.

**Comment by simosacchi** [comment]
@mcanouil  I opened the GitHub repo for the project and prepared a GitHub Pages version as well, so the full source is available:
- repo: https://github.com/EUI-Library/research-data-guide
- page: https://eui-library.github.io/research-data-guide/

In terms of workflow, I am running Quarto 1.5.43 within VScode 1.90.0 on MacOS 14.4.1 

_Please note that the behaviour started happening more or less a week ago. I have only used Quarto for a few weeks now and it used to work properly. I am not aware on any change and/or update in the system in the meanwhile._

### Reproducing the issue
Here is how (at least) I ended up with the WARN issue:
- checkout the repo
- run `quarto preview`
  - server goes up as it should and page is served in the browser
  - if you try to navigate to a different page a "WARN: connection closed before message completed" message is sent to STDOUT and the original page is reloaded twice (at least according to STDOUT)
  - if you open a page in a different browser tab, the page opens, but it triggers a reloading loop on both open tabs (according to STDOUT and the browser icon)
  - loop stops as you close one of the two tabs

The only way I found to avoid the behaviour above is to:
- checkout the repo
- launch preview with render option `quarto preview --render html`
- navigation works perfectly, no issue
- now, edit a page (e.g. dmp.qmd)
  - page is reloaded (according to watch input), and edit is rendered
  - if you try to navigate to a different page, a "WARN: connection closed before message completed" message is sent to STDOUT and the original page is reloaded twice (at least according to STDOUT)
 
I hope the above can be considered a reproducible example. At least, that is the behaviour on my machine (I have the same behaviour with Chrome, Safari and Firefox).

Thank you for any help!

**Reply by mcanouil** [reply]
Thanks for the details. We'll get back to you.

> Please note that the behaviour started happening more or less a week ago.

Seems to coincide with the monthly VSCode update (first Thursday of the month).

**Reply by simosacchi** [reply]
On the VSCode update timing: even running commands in the terminal (not inside VSCode) the issue persists.

Looking forward to hearing back. Thanks!

**Reply by gordonwoodhull** [reply]
FWIW I see slightly different, but also bad, behavior on cloning your repo and previewing (via the command line).

On my machine the navigation does go through, but the page reloads multiple times, and the console shows both the previous page and the new page load many times.

Behavior is consistent with rendering somehow touching a source file, causing a re-render. Presumably the warning happens when the re-render interrupts an HTTP transfer, but the reload does not always cause the warning.

Looking at the timestamps in the directory, I see the `.quarto` directory getting updated every reload, but I don't see changes to other files.

**Reply by simosacchi** [reply]
thanks @gordonwoodhull . I feared that there may be differences when reproducing, as the behaviour I am experiencing seems so odd.

**Comment by skwde** [comment]
The same also happens to me for some time now (~O(weeks)) when previewing revealjs slides.

Even though in terminal it looks normal and reloads, the VSCode preview (Simple browser) is not updated.
When I manually click the reload button in it, the warning is displayed multiple times. 
The number of the same warning though varies.

Note, there is a possibly related (closed by @cscheid on Nov 2022) issue: https://github.com/quarto-dev/quarto-cli/issues/3045

**Reply by andrewpbray** [reply]
I've been bumping into this issue for a couple weeks now as well. Happy to provide reprex if that's helpful.

**Reply by cscheid** [reply]
> Happy to provide reprex if that's helpful.

Yes, please and thank you.

**Reply by andrewpbray** [reply]
I'm realizing a minimal reprex will be challenging to make. I haven't encountered these problems with `quarto preview` until recently and only in one project, a [very complex one](https://github.com/berkeley-stat20/fall-2024) with lots of extensions and pre-render scripting that runs in a [container](https://github.com/stat20/stat20-docker).

Might be easier to start with a screen capture of what the behavior looks like in RStudio.

https://www.dropbox.com/scl/fi/k3gl92yv0snnbj0k7f198/quarto-preview-hang.mov?rlkey=bd4tbjk7w88zoavor9b79t92a&dl=0

Relevant behavior:

1. On previewing the lab qmd file (one of ~150 qmds in this project), it behaves as expected
2. From localhost in the browser, if I click the link to the index, it seems to time out and remains on the lab page.
3. If you note the reload indicator in the browser, this kicks off an unusual continuous re-rendering of this page
4. You can override the timeout by entering the URL directly into the browser.
5. Once on the index, the continuous re-rendering behavior is still going.

Not shown: the background job tab in RStudio doesn't respond to the STOP button and I have to force quit RStudio at this point.

How should I got about troubleshooting this?

This running Quarto v1.6.9 on MacOS Sonoma 14.6.1 and Chrome 128.0.6613.115.

**Reply by cscheid** [reply]
Before we have a small reprex to be able to localize the problem, it'll be hard for us to do much about it :(

**Reply by andrewpbray** [reply]
Understood. I can't even put a bound on how much time it would take me to snip away complexity to isolate the behavior, so I don't think I'll be able to do this (and I certainly don't expect you to!).

In the meantime, is there any logging that I can check to try to troubleshoot myself?

**Reply by cscheid** [reply]
I can't think of anything, unfortunately.

**Reply by gordonwoodhull** [reply]
It’s easy to repro using the original @simosacchi repository, which is fairly straightforward but has a lot of content.

**Reply by andrew-1234** [reply]
> Before we have a small reprex to be able to localize the problem, it'll be hard for us to do much about it :(

reprex was provided just one comment above what you replied to

**Reply by cscheid** [reply]
@andrew-1234 "small reprex" with emphasis on _small_, FYI.

**Reply by winniehell** [reply]
size of the document may be the key here. preview worked for me without problems but as my document grew I'm now getting that error message on almost every run. involved software versions have not changed in the meantime.

so providing a small reproduction example may not be possible.

**Reply by FelixBenning** [reply]
I stumbled into this error creating a book template, so by definition of a template this a fairly small reproduction: https://github.com/FelixBenning/math-book/tree/bbd2065b6cd7d701b5b7383dc819f9fbd868c68d

The largest part in this template is probably the inclusion of the [h5p standalone](https://github.com/tunapanda/h5p-standalone) project in `extension/assets/`.

Notes for reproduction:
I am using 
```
quarto preview
```
in the terminal. And my default browser that opens the preview is Firefox Version 139.0.1

I get
```
Watching files for changes
Browse at http://localhost:5401/
GET: /
GET: /
GET: /intro.html
WARN: connection closed before message completed
GET: /
GET: /intro.html
WARN: connection closed before message completed
GET: /
GET: /intro.html
WARN: connection closed before message completed
GET: /
GET: /intro.html
WARN: connection closed before message completed
```

If I use the vscode integrated preview, the warning messages are not displayed but the window still seems to be reloading too much.

**Comment by benkeks** [comment]
The whole navigation of previews seems to be fundamentally broken if book chapters have long render times. (This issue has also annoyed me a lot over the last months of writing my thesis...)

The problem seems to be connected to the `window.location.reload()` happening in

https://github.com/quarto-dev/quarto-cli/blob/a496e0e0e7f02f23b95ed678a5580cce34f4ed3f/src/webui/quarto-preview/src/server/navigation.ts#L26-L30

If I change the `else` case to only fire if a `target` has been set, the previewer can handle navigation on documents with long render times.

```ts
if (target && (target !== normalizeTarget(window.location.pathname))) {
  window.location.replace(target);
} else if (target) {
  window.location.reload();
}
```

However, I have a really hard time to understand how the previewer messages are supposed to interact and what role the reload should play in there if `target` is the empty string. So I can't propose a PR. (Removing the window reload creates another glitch where the “Render” dialog does not go away after the rendering, if one accidentally starts multiple renders, e.g. due to mousehover Tippy.js popups. At least for me, the latter glitch is far less annoying than perpetually using `touch _quarto.yml`...)

Related to issue #11475.

**Comment by joshbrake** [comment]
Just wanted to chime in to say that I too am experiencing this issue and agree with @winniehell that the size of the document seems to be a key variable. In my experience, the preview works fine until the project grows to a certain size after which this error happens constantly and makes the preview server functionality very frustrating to attempt to use.

I am suspicious that @benkeks is on the right track with long render times being the culprit.

**Comment by k-soomro** [comment]
Was this issue ever resolved? I am experiencing the same, but randomly. It just happens sometimes and sometimes doesn't.
